File: assignments/assignment-1-solution/src/CrazyEights/Player/HumanPlayer.cs
using CrazyEights.Cards;
using CrazyEights.Domain;
using CrazyEights.Game;

namespace CrazyEights.Player;

public class HumanPlayer : PlayerBase
{
    public const string DefaultName = "Player";

    private const bool ShowMyHand = true;

    public HumanPlayer(string name, Hand hand) : base(name, hand, ShowMyHand)
    {
    }

    public override ICard SelectCard(TurnContext turnContext)
    {
        var playableCards =
            Hand.GetPlayableCards(turnContext.CurrentSuit, turnContext.CurrentRank);

        //cannot select a card
        if (playableCards.Count == 0)
            return UnselectedCard.Instance;

        ShowPlayableCardsInHand(turnContext, playableCards);

        var isValidInput = false;
        ICard selectedCard = UnselectedCard.Instance;
        while (!isValidInput)
        {
            var input = GameConsole.ReadLine("Choose a card number to play or 0 to draw: ");
            var isNumber = int.TryParse(input, out var selectedNumber);
            if (!isNumber || selectedNumber < 0 || selectedNumber > playableCards.Count)
            {
                GameConsole.WriteLine("- Invalid choice! Please try again.");
                continue;
            }

            if (selectedNumber > 0)
                selectedCard = playableCards[selectedNumber - 1].Card;
            isValidInput = true;
        }

        return selectedCard;
    }

    private static void ShowPlayableCardsInHand(TurnContext turnContext, IReadOnlyList<PlayableCard> playableCards)
    {
        var currentPlayer = turnContext.CurrentPlayer;
        var name = currentPlayer.Name;

        GameConsole.WriteLine();
        GameConsole.WriteLine($"{name}'s playable cards");

        for (var i = 0; i < playableCards.Count; i++)
        {
            var playableCard = playableCards[i];
            var card = playableCard.Card;
            var selector = i + 1;
            var cardDescription = card.GetDescription();
            var playableReason = playableCard.PlayableReason;
            GameConsole.WriteLine($"  [{selector}] {cardDescription} ({playableReason})");
        }
    }

    public override SuitType SelectSuit(GameContext gameContext, TurnContext turnContext)
    {
        var currentSuit = gameContext.DiscardPile.ActiveSuit;
        var suits = Suit.GetSuits();
        var name = turnContext.CurrentPlayer.Name;

        //select a suit
        var isValidInput = false;
        var selectedSuit = SuitType.NotSet;

        //choose the first letter of a valid suit
        while (!isValidInput)
        {
            GameConsole.WriteLine();
            GameConsole.WriteLine($"{name}, you played a wildcard! Choose a suit:");
            foreach (var suit in suits)
            {
                var suitName = Suit.GetSuitName(suit);
                var letter = suitName[0].ToString().ToUpper();
                var isCurrentSuitNote = currentSuit == suit ? " (current suit)" : "";
                GameConsole.WriteLine($"  [{letter}] {suit}{isCurrentSuitNote}");
            }

            var input = GameConsole.ReadLine("Enter the letter of your chosen suit: ");
            if (string.IsNullOrWhiteSpace(input) || input.Length != 1)
            {
                GameConsole.WriteLine("- Invalid choice! Please try again.");
                continue;
            }

            var selectedLetter = char.ToUpper(input[0]);
            switch (selectedLetter)
            {
                case 'H':
                    selectedSuit = SuitType.Hearts;
                    break;
                case 'D':
                    selectedSuit = SuitType.Diamonds;
                    break;
                case 'C':
                    selectedSuit = SuitType.Clubs;
                    break;
                case 'S':
                    selectedSuit = SuitType.Spades;
                    break;
                default:
                    GameConsole.WriteLine("- Invalid choice! Please try again.");
                    continue;
            }

            isValidInput = true;
        }

        return selectedSuit;
    }

    public override bool WillPlayDrawnCard(TurnContext turnContext, ICard drawnCard)
    {
        GameConsole.WriteLine();
        var playDrawnCard = GameConsole.PromptYesNo("You drew a playable card. Play it now (Y/N): ");
        return playDrawnCard;
    }
}

File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/player/HumanPlayer.java
package edu.kennesaw.crazy8s.player;

import edu.kennesaw.crazy8s.cards.Card;
import edu.kennesaw.crazy8s.cards.UnselectedCard;
import edu.kennesaw.crazy8s.domain.PlayableCard;
import edu.kennesaw.crazy8s.domain.Suit;
import edu.kennesaw.crazy8s.domain.SuitType;
import edu.kennesaw.crazy8s.game.GameConsole;
import edu.kennesaw.crazy8s.game.GameContext;
import edu.kennesaw.crazy8s.game.TurnContext;
import java.util.List;

/**
 * Human-controlled player that chooses cards via console input.
 */
public class HumanPlayer extends PlayerBase {
    public static final String DEFAULT_NAME = "Player";

    private static final boolean SHOW_MY_HAND = true;

    public HumanPlayer(String name, Hand hand) {
        super(name, hand, SHOW_MY_HAND);
    }

    @Override
    public Card selectCard(TurnContext turnContext) {
        List<PlayableCard> playableCards =
                getHand().getPlayableCards(turnContext.getCurrentSuit(), turnContext.getCurrentRank());

        if (playableCards.isEmpty()) {
            return UnselectedCard.getInstance();
        }

        showPlayableCardsInHand(turnContext, playableCards);

        boolean isValidInput = false;
        Card selectedCard = UnselectedCard.getInstance();
        while (!isValidInput) {
            String input = GameConsole.readLine("Choose a card number to play: ");
            Integer selectedNumber = tryParseInt(input);
            if (selectedNumber == null || selectedNumber < 1 || selectedNumber > playableCards.size()) {
                GameConsole.writeLine("- Invalid choice! Please try again.");
                continue;
            }

            selectedCard = playableCards.get(selectedNumber - 1).getCard();
            isValidInput = true;
        }

        return selectedCard;
    }

    private static void showPlayableCardsInHand(TurnContext turnContext, List<PlayableCard> playableCards) {
        Player currentPlayer = turnContext.getCurrentPlayer();
        String name = currentPlayer.getName();

        GameConsole.writeLine();
        GameConsole.writeLine(name + "'s playable cards");

        for (int i = 0; i < playableCards.size(); i++) {
            PlayableCard playableCard = playableCards.get(i);
            Card card = playableCard.getCard();
            int selector = i + 1;
            String cardDescription = card.getDescription();
            String playableReason = playableCard.getPlayableReason();
            GameConsole.writeLine(String.format("  [%d] %s (%s)", selector, cardDescription, playableReason));
        }
    }

    @Override
    public SuitType selectSuit(GameContext gameContext, TurnContext turnContext) {
        SuitType currentSuit = gameContext.getDiscardPile().getActiveSuit();
        List<SuitType> suits = Suit.getSuits();
        String name = turnContext.getCurrentPlayer().getName();

        boolean isValidInput = false;
        SuitType selectedSuit = SuitType.NOT_SET;

        while (!isValidInput) {
            GameConsole.writeLine();
            GameConsole.writeLine(name + ", you played a wildcard! Choose a suit:");
            for (SuitType suit : suits) {
                String suitName = Suit.getSuitName(suit);
                String letter = suitName.substring(0, 1).toUpperCase();
                String isCurrentSuitNote = currentSuit == suit ? " (current suit)" : "";
                GameConsole.writeLine(String.format("  [%s] %s%s", letter, suit, isCurrentSuitNote));
            }

            String input = GameConsole.readLine("Enter the letter of your chosen suit: ");
            if (input == null || input.length() != 1) {
                GameConsole.writeLine("- Invalid choice! Please try again.");
                continue;
            }

            char selectedLetter = Character.toUpperCase(input.charAt(0));
            switch (selectedLetter) {
                case 'H':
                    selectedSuit = SuitType.HEARTS;
                    break;
                case 'D':
                    selectedSuit = SuitType.DIAMONDS;
                    break;
                case 'C':
                    selectedSuit = SuitType.CLUBS;
                    break;
                case 'S':
                    selectedSuit = SuitType.SPADES;
                    break;
                default:
                    GameConsole.writeLine("- Invalid choice! Please try again.");
                    continue;
            }

            isValidInput = true;
        }

        return selectedSuit;
    }

    @Override
    public boolean willPlayDrawnCard(TurnContext turnContext, Card drawnCard) {
        GameConsole.writeLine();
        return GameConsole.promptYesNo("You drew a playable card. Play it now (Y/N): ");
    }

    private static Integer tryParseInt(String value) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException ex) {
            return null;
        }
    }
}
